examples/OcclusionSaliency.ipynb

```diff
@@ -19,81 +16,101 @@
-    "OF SUCH DAMAGE.\n",
-    "\"\"\"\n",
-    "from xaitk_saliency.impls.perturb_image.sliding_window import SlidingWindowPerturb\n",
-    "from xaitk_saliency.impls.vis_sal_classifier.occlusion_scoring import OcclusionScoring\n",
-    "\n",
-    "import matplotlib.pyplot as plt\n",
-    "from PIL import Image\n",
-    "from torchvision import transforms\n",
-    "from torch.autograd import Variable\n",
-    "import torchvision.models as models\n",
-    "import torch\n",
+    "License for test image used may be found in 'catdog-LICENSE.txt'."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "(1000,)"
+      ]
+     },
+     "execution_count": 1,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Set up our \"black box\" classifier using PyTorch and it's ImageNet pretrained ResNet18.\n",
     "import numpy as np\n",
-    "from operator import itemgetter\n",
     "import os\n",
+    "import PIL.Image\n",
+    "import torch\n",
+    "from torch.autograd import Variable\n",
+    "import torchvision.models as models\n",
+    "import torchvision.transforms as transforms\n",
     "\n",
+    "CUDA_AVAILABLE = torch.cuda.is_available()\n",
     "\n",
-    "# Test image to be explained\n",
-    "test_image = 'catdog.png'\n",
-    "# Imagenet class for which saliency maps need to be computed \n",
-    "sal_classes = ['boxer', 'tiger cat']\n",
-    "\n",
-    "# Constructing xaitk implementations for OcclusionBased Saliency genration with\n",
-    "# sliding windows.\n",
-    "slid_algo = SlidingWindowPerturb(window_size=(24, 24), stride=(5, 5))\n",
-    "occ_alg = OcclusionScoring()\n",
     "\n",
-    "# Creating an example deep feature extractor to describe images\n",
     "model = models.resnet18(pretrained=True)\n",
     "model = model.eval()\n",
-    "model = model.cuda()\n",
+    "if CUDA_AVAILABLE:\n",
+    "    model = model.cuda()\n",
+    "\n",
+    "# These are some simple helper functions to perform prediction with this model\n",
+    "model_input_size = 224\n",
+    "model_loader = transforms.Compose([\n",
+    "    transforms.Resize(model_input_size), \n",
+    "    transforms.ToTensor(),\n",
+    "    transforms.Normalize(\n",
+    "        mean=[0.485, 0.456, 0.406],\n",
+    "        std=[0.229, 0.224, 0.225]\n",
+    "    ),\n",
+    "])\n",
+    "\n",
+    "\n",
+    "def image_loader(image):\n",
+    "    image = model_loader(image).float()\n",
+    "    if CUDA_AVAILABLE:\n",
+    "        image = image.cuda()\n",
+    "    return image.unsqueeze(0)\n",
+    "\n",
+    "\n",
+    "def predict_with_model(model, image_tensor):\n",
+    "    featureVec = model(image_tensor)\n",
+    "    # Converting feature extractor output to probabilities.\n",
+    "    return torch.nn.functional.softmax(featureVec, dim=1).cpu().detach().numpy().squeeze()\n",
```

Comment:
Minor point: Is the detach necessary here? Since we run the model in eval mode, I don't think gradients are computed.

Reply:
Apparently it is. I tried removing it but for some reason I have yet to understand the output tensor is marked for backprop, despite tests ensuring the input is a variable not for grad and that the network is indeed in eval mode.

Reply:
Weird, I can't argue with that then...